### ISSUE TITLE
systemd does not recognize disks [ARM] [Systemd v.215]

### ISSUE THREAD

**Issue by 0x54726F6E** [systemd/systemd#2257]
Hello there,

I have been trying to get systemd to mount a luks storage device on an armv6 (rasperry pi).
It is working with the legacy setup in debian, but not with the new systemd which I think is a beatiful (if pretty blackbox) system in my opinion. :)

So I got the kernel running and loading the initrd I had previously generated with dracut and as I needed a pretty generic way I devided to go for the cmdline option to mount my devices:

root=UUID=b6756e74-59c5-4a19-af44-d7610d1a2b7d rd.luks.uuid=2089b2d1-e607-4ea0-8aed-29d08b066cbd systemd.log_target=kmsg systemd.log_level=debug rd.udev.log-priority=debug rd.luks=1

luks etc are included and the UUID's are correct, so not the standard kind of error here. (see attached logfile)

After some time the system times out for luks target and root target (which is located in the luks container) and I started debugging why that is: The interesting thing is - running cryptsetup by hand and mounting rootfs is working. So I set a symlink to /dev/root and ran "exit". Instead of continuing to boot, it sais "not all disks have been found" and keeps running the job for disk recognition for ever.

To make it some more interesting: The /dev/disk/by-uuid/ files both exists and work when used for mounting. Almost looks like systemd is just never notified that they exists. I even tried to run the dev-disk-by-uuidXXXX.device service by hand and it still says "expecting disk uuid ... ".

The system I used to generate the dracut image has included crypt etc. and I have only left out kernel drivers etc. as they are included in the kernel. (dracut --no-kernel)
There is no fstab nor crypttab as the only used disks are directed by cmdline.

Thanks for looking into it, I have been at is for 3 days and can't really find any mention of the problem on the internet at all. :/

Keep up the good work!

0x54726F6E

Attachments: 
rdsosreport.txt 
journalctl.txt - output from journalctl -a
[journalctl.txt](https://github.com/systemd/systemd/files/76712/journalctl.txt)
[rdsosreport.txt](https://github.com/systemd/systemd/files/76713/rdsosreport.txt)

additional information:

systemd recognizes the device if I start the .device by running systemctl, put it in background and run a daemon-reload. but due to all the timers having run out already does not continue to boot. Maybe unrelated, but systemd complains that the luks unitfile has changed on disk and recommends running the daemon-reload.

**Comment by poettering** [comment]
Does "udevadm info /dev/..." show the "systemd" tag on the block device you want to boot from? Unless "systemd" appears in the TAGS= line of the block device, systemd won#t notice the device. If the tag is missing this indicates some DM issues, maybe it was built lacking proper udev support? 

Which distro is this? Which systemd version is this?

**Comment by 0x54726F6E** [comment]
yes is does, should have mentioned it before. TAGS Line shows ":systemd:".

Systemd Version is 215 and its debian jessie armel but with a custom built kernel v 4.1.6 from the rpi repos.
The dracut version string appears to be empty when I run it, but the package version is: 040+1-1

I tried this before with a custom initrd and it worked, but systemd wouldn't recognize the already mounted devices. So I switched to dracut as it already has systemd with udev inside the initrd.

EDIT: There is also a udev rule for the device that includes the command to execute the crypt-disks but as I ran the command by hand I got no output and the return code 0. I also looked into unit dependency problems, but they appear to be fine according to the image I have attached.

(dang, svg not supported. Get it at www.ruehenbeck.org/systemd/systemd.svg and www.ruehenbeck.org/systemd/dotgraph.dot )

**Comment by poettering** [comment]
Sorry, we do not track issues with such old systemd versions upstream. Please ask your distro for help on this. Please only report issues with the two most recent systemd versions upstream, thanks!

**Comment by 0x54726F6E** [comment]
As I solved it, I will leave some information about how I did it here: 

As I did not find anything in the logs about the missing kernel modules that were most likely flagged as "optional" here: http://cgit.freedesktop.org/systemd/systemd/tree/README
and are not really optional, I would suggest you add an informing message about missing kernel modules. As you recommended I used dracut as "everything is already there"... and sadly spent 3 months chasing the error in systemd (event in the most recent version btw.) except in my kernel options. (as everything was working "by hand" only systemd did not pick it up.)

I added the following kernel options to device detection working with Linux 4.1.15 on arm:

+CONFIG_CFS_BANDWIDTH=y
+CONFIG_CHECKPOINT_RESTORE=y
+CONFIG_SHMEM=y
+CONFIG_BCM2835_SMI=y
+CONFIG_SCSI_PROC_FS=y
+CONFIG_SCSI_SCAN_ASYNC=y
+CONFIG_BCM2835_SMI_DEV=y
+CONFIG_TMPFS=y
+CONFIG_TMPFS_POSIX_ACL=y
+CONFIG_TMPFS_XATTR=y

Fair enough some of them are listed as "optional but highly recommended".

Finally as I hear you get quite some greavance about systemd I want to thank you for all the work you have been putting into this. Keep it up, and if we ever meet in person I will buy you a beer. ;)
